src/tools/pkcs11-tool.c

```diff
@@ -3469,64 +3521,99 @@ unwrap_key(CK_SESSION_HANDLE session)
+		key_type = CKK_GENERIC_SECRET;
+	} else if (strncasecmp(opt_key_type, "RSA:", strlen("RSA:")) == 0) {
+		length = "0"; // No key length for RSA keys
+		key_type = CKK_RSA;
+		class = CKO_PRIVATE_KEY;
+	} else if (strncasecmp(opt_key_type, "EC:", strlen("EC:")) == 0) {
+		length = "0"; // No key length for EC keys
+		key_type = CKK_EC;
+		class = CKO_PRIVATE_KEY;
+	} else {
+		util_fatal("Unsupported key type %s", opt_key_type);
+	}
 
-		FILL_ATTR(keyTemplate[n_attr], CKA_KEY_TYPE, &key_type, sizeof(key_type));
-		n_attr++;
+	FILL_ATTR(keyTemplate[n_attr], CKA_KEY_TYPE, &key_type, sizeof(key_type));
+	n_attr++;
 
-		if (opt_is_sensitive != 0) {
-			FILL_ATTR(keyTemplate[n_attr], CKA_SENSITIVE, &_true, sizeof(_true));
-			n_attr++;
-		} else {
-			FILL_ATTR(keyTemplate[n_attr], CKA_SENSITIVE, &_false, sizeof(_false));
-			n_attr++;
-		}
+	if (opt_is_sensitive != 0) {
+		FILL_ATTR(keyTemplate[n_attr], CKA_SENSITIVE, &_true, sizeof(_true));
+	} else {
+		FILL_ATTR(keyTemplate[n_attr], CKA_SENSITIVE, &_false, sizeof(_false));
+	}
+	n_attr++;
 
-		if (opt_key_usage_default || opt_key_usage_decrypt) {
+	if (opt_key_usage_default || opt_key_usage_decrypt) {
+		if (class != CKO_PRIVATE_KEY) {
 			FILL_ATTR(keyTemplate[n_attr], CKA_ENCRYPT, &_true, sizeof(_true));
 			n_attr++;
-			FILL_ATTR(keyTemplate[n_attr], CKA_DECRYPT, &_true, sizeof(_true));
-			n_attr++;
 		}
-		if (opt_key_usage_wrap) {
+		FILL_ATTR(keyTemplate[n_attr], CKA_DECRYPT, &_true, sizeof(_true));
+		n_attr++;
+	}
+	if (opt_key_usage_wrap) {
+		if (class != CKO_PRIVATE_KEY) {
 			FILL_ATTR(keyTemplate[n_attr], CKA_WRAP, &_true, sizeof(_true));
 			n_attr++;
-			FILL_ATTR(keyTemplate[n_attr], CKA_UNWRAP, &_true, sizeof(_true));
-			n_attr++;
-		}
-
-		if (opt_is_extractable != 0) {
-			FILL_ATTR(keyTemplate[n_attr], CKA_EXTRACTABLE, &_true, sizeof(_true));
-			n_attr++;
-		} else {
-			FILL_ATTR(keyTemplate[n_attr], CKA_EXTRACTABLE, &_false, sizeof(_true));
-			n_attr++;
 		}
-		/* softhsm2 does not allow to attribute CKA_VALUE_LEN, but MyEID card must have this attribute
-                   specified. We set CKA_VALUE_LEN only if the user sets it in the key specification. */
-		key_length = (unsigned long)atol(length);
-		if (key_length != 0) {
-			FILL_ATTR(keyTemplate[n_attr], CKA_VALUE_LEN, &key_length, sizeof(key_length));
+		FILL_ATTR(keyTemplate[n_attr], CKA_UNWRAP, &_true, sizeof(_true));
+		n_attr++;
+	}
+	if (opt_key_usage_sign) {
+		if (class != CKO_PRIVATE_KEY) {
+			FILL_ATTR(keyTemplate[n_attr], CKA_VERIFY, &_true, sizeof(_true));
 			n_attr++;
 		}
+		FILL_ATTR(keyTemplate[n_attr], CKA_SIGN, &_true, sizeof(_true));
+		n_attr++;
+	}
+
+	if (opt_is_extractable != 0) {
+		FILL_ATTR(keyTemplate[n_attr], CKA_EXTRACTABLE, &_true, sizeof(_true));
+	} else {
+		FILL_ATTR(keyTemplate[n_attr], CKA_EXTRACTABLE, &_false, sizeof(_true));
+	}
+	n_attr++;
+
+	/* softhsm2 does not allow to attribute CKA_VALUE_LEN, but MyEID card must have this attribute
```

Comment:
I'm conflicted regarding this approach. The PKCS11 spec explicitly mentions that CKA_VALUE_LEN **must** be absent from the template when unwrapping an AES key. Do MyEID cards not comply to PKCS11?

It should at least be mentioned somewhere that a PKCS11 module will return an error if anything other than `--type AES:` is present in an AES key unwrap command.

Reply:
@popovec can you have a look to this comment regarding the myeid driver and the need for the CKA_VALUE_LEN for AES keys?

Reply:
First, the technical reasons why MyEID needs to know the length of the unwrapped key before the operation itself - before the "unwrap" operation, MyEID needs to create a file in which the corresponding AES key is stored. The length of the file must correspond to the length of the AES key.

In a certain way, pkcs#11 (ver.2.40, 2.8.5 AES-CBC or ver.3.0, 2.10.5 AES-CBC) explains it, where, as far as I understand, CKA_VALUE_LEN can also be part of template:

"For unwrapping, the mechanism decrypts the wrapped key, and truncates the result according to the
CKA_KEY_TYPE attribute of the template and, if it has one, and the key type supports it, the
CKA_VALUE_LEN attribute of the template. The mechanism contributes the result as the CKA_VALUE
attribute of the new key; other attributes required by the key type must be specified in the template"

The description of pkcs#11 is a bit imprecise for me, the table in the relevant pkcs#11 section shows it more clearly: C_UnwrapKey -> Output length -> determined by type of key being unwrapped or CKA_VALUE_LEN. 

From this I conclude that the implementation can use CKA_VALUE_LEN specified in template.

Reply:
The quote is correct, however AES keys in fact do not support it. In [ver.3.0, 2.10.2 AES secret key objects]/[ver.2.40, 2.8.2 AES secret key objects] the CKA_VALUE_LEN attribute has a footnote that specifies that for AES keys it must be absent from the template when unwrapping.

![image](https://github.com/OpenSC/OpenSC/assets/147070003/2a0a9de9-b97f-4997-9283-5e5f6764a273)
![image](https://github.com/OpenSC/OpenSC/assets/147070003/8615b71c-43eb-4ccd-95a2-c3fceaaa80bc)

Reply:
As I already mentioned, the pkcs#11 specification is sometimes rather difficult to read for me, it is difficult for me to find out which case applies to what. I am looking at the document: http://docs.oasis-open.org/pkcs11/pkcs11-curr/v3.0/pkcs11-curr-v3.0.pdf and specifically on the AES-CBC method (or AES-ECB, which MyEID also has implemented). The tables for the mentioned methods (table 71 and 72) do not mention footnotes for CKA_VALUE_LEN, (in contrast to table 70, which contains the footnotes you mentioned). For this reason, I believe that the AES-CBC/AES_ECB __method__ (used for unwrap operation) already allows the use of CKA_VALUE_LEN while "AES_secret key" __objects__ do not allow the use of CKA_VALUE_LEN.

Reply:
The spec is indeed difficult to read.
My understanding is that when unwrapping a key whose key type supports CKA_VALUE_LEN in the unwrapping template, AES-ECB/AES-CBC will make use of this attribute and otherwise will only use the CKA_KEY_TYPE attribute (as well as the length of the wrapped key) to determine the length of the key, hence
> "the mechanism [...] truncates the result according to [...], if it has one _(e.g. in the template)_ and the key type supports it, the CKA_VALUE_LEN attribute of the template."
AES key objects have a CKA_VALUE_LEN attribute **but** do not support specifying it in the unwrapping template.

Reply:
I think OpenSC needs to work in "benevolent" mode and occasionally allow things not allowed in the spec. Currently, the pkcs11-tool allows the entry of the AES key type without specifying the length, or with the entry of the key length, and accordingly the corresponding field CKA_VALUE_LEN in the template will be filled or not filled. If one day it turns out that this behavior is not allowed, I will try to modify the unwrap function for the MyEID driver so that this driver also does not need to know the length of the unwrapped key in advance.

Reply:
I completely agree, I never meant to prevent specifying the length altogether.
I'll just write a note in the documentation to make the user know that in some cases, they will need to discard the length in the key type.

Reply:
I have been trying to follow this discussion. 

> First, the technical reasons why MyEID needs to know the length of the unwrapped key before the operation itself - before the "unwrap" operation, MyEID needs to create a file in which the corresponding AES key is stored. The length of the file must correspond to the length of the AES key.

As pointed out PKCS11 table 11 (6)  "6 MUST not be specified when object is unwrapped with C_UnwrapKey."

But if the MyEID is trying to unwrap an AES key, PKCS11 3.0 current says:
```
2960 For unwrapping, the mechanism decrypts the wrapped key, and truncates the result according to the 
2961 CKA_KEY_TYPE attribute of the template and, if it has one, and the key type supports it, the 
2962 CKA_VALUE_LEN attribute of the template. The mechanism contributes the result as the CKA_VALUE 
2963 attribute of the new key; other attributes required by the key type must be specified in the template.
```

The encrypted length of an encrypted AES key is the same size as the unencrypted form, as they are all multiples of the blocksize.  So in this case the encrypted size would need to present in the  `CK_UnwrapKey` `CK_ULONG ulWrappedKeyLen` and the  `CK_ATTRIBUTE_PTR pTemplate` would contain `CK_KEY_TYPE keyType = CKK_AES`,
`CK_VOID_PTR pValue;` and `CK_ULONG ulValueLen;`

which could then be use to in place of  the "MUST not be present" `CKA_VALUE_LEN`" in question.  

The `ulWrappedKeyLen` would be the same as the The resulting `ulValueLen`  in `CKA_VALUE`  `CK_ATTRIBUTE` 
```
CK_ATTRIBUTE_TYPE type;
CK_VOID_PTR pValue; 
CK_ULONG ulValueLen;
```
So in effect MyEID  "know**S** the length of the unwrapped key in advance."
Did I get this right?

Reply:
> The encrypted length of an encrypted AES key is the same size as the unencrypted form

Sadly this is rarely true... RSA wrapping mechanisms, PKCS#7 padded AES wrapping mechanisms, AES GCM (among others) will always output a wrapped key that is longer than the initial key.

Reply:
> So in effect MyEID "knowS the length of the unwrapped key in advance."
> Did I get this right?

Yes, you are right, with the AES-CBC and AES-ECB mechanisms it is probably still possible to determine the length of the resulting key from the length of the encrypted data. Padding does not apply here. This of course means that AES-CBC/AES-ECB does not allow wrapping a 192-bit key using a 256-bit key or a 128-bit key either, as the resulting key does not have a length in multiples of the block size, so AES-CBC-PAD must be used in such cases.

For cases where it is possible to derive the length of the resulting key without CKA_VALUE_LEN information, it will probably be appropriate to correct the unwrapping code so that it derives the relevant key length from the input data.

Reply:
The object into which the key is unwrapped is created in the function CK_RV C_UnwrapKey(), specially by calling 
`rv = sc_create_object_int(hSession, pTemplate, ulAttributeCount, phKey, 0);` . At this point, CKA_VALUE_LEN must be present in the template, so "adding" CKA_VALUE_LEN to the template for specific cases should happen before this call.